ocb3/src/lib.rs

```diff
@@ -0,0 +1,588 @@
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg"
+)]
+#![deny(unsafe_code)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+use core::marker::PhantomData;
+
+pub use aead::{
+    self, generic_array::GenericArray, AeadCore, AeadInPlace, Error, KeyInit, KeySizeUser,
+};
+use cipher::{
+    consts::{U0, U12, U16},
+    BlockDecrypt, BlockEncrypt, BlockSizeUser,
+};
+use subtle::ConstantTimeEq;
+
+/// Constants used, reexported for convenience.
+pub mod consts {
+    pub use cipher::consts::{U0, U12, U16};
+}
+
+mod util;
+
+use crate::util::{double, inplace_xor, ntz, Block};
+
+/// Number of L values to be precomputed. Precomputing m values, allows
+/// processing inputs of length up to 2^m blocks (2^m * 16 bytes) without
+/// needing to calculate L values at runtime.
+///
+/// By setting this to 32, we can process inputs of length up to 1 terabyte.
+#[cfg(target_pointer_width = "64")]
+const L_TABLE_SIZE: usize = 32;
+
+/// Number of L values to be precomputed. Precomputing m values, allows
+/// processing inputs of length up to 2^m blocks (2^m * 16 bytes) without
+/// needing to calculate L values at runtime.
+#[cfg(target_pointer_width = "32")]
+const L_TABLE_SIZE: usize = 16;
+
+/// Max associated data.
+pub const A_MAX: usize = 1 << (L_TABLE_SIZE + 4);
+/// Max plaintext.
+pub const P_MAX: usize = 1 << (L_TABLE_SIZE + 4);
+/// Max ciphertext.
+pub const C_MAX: usize = 1 << (L_TABLE_SIZE + 4);
+
+/// OCB3 nonce
+pub type Nonce<NonceSize> = GenericArray<u8, NonceSize>;
+
+/// OCB3 tag
+pub type Tag<TagSize> = GenericArray<u8, TagSize>;
+
+/// Trait implemented for valid tag sizes
+pub trait TagSize: private::SealedTagSize {}
+impl<T: private::SealedTagSize> TagSize for T {}
+/// Trait implemented for valid nonce sizes
+pub trait NonceSize: private::SealedNonceSize {}
+impl<T: private::SealedNonceSize> NonceSize for T {}
+
+// Adapted from https://github.com/rustcrypto/AEADs/blob/2209bcaa9edc65e9a60498e7ece5b50e66f32ebf/aes-gcm/src/lib.rs#L143-L157
+mod private {
+    use aead::generic_array::ArrayLength;
+    use cipher::{consts, Unsigned};
+
+    // Sealed traits stop other crates from implementing any traits that use it.
+    pub trait SealedTagSize: ArrayLength<u8> + Unsigned {}
+    pub trait SealedNonceSize: ArrayLength<u8> + Unsigned {}
+
+    // Tags are <= 128 bits
+    impl SealedTagSize for consts::U8 {}
+    impl SealedTagSize for consts::U12 {}
+    impl SealedTagSize for consts::U16 {}
+
+    // Nonces are <= 120 bits
+    impl SealedNonceSize for consts::U12 {}
+}
+
+/// AES-OCB3: generic over an AES implementation, nonce size, and tag size.
+///
+/// WARNING: Unless absolutely necessary, prefer the aliases Aes128Ocb3 and
+/// Aes256Ocb3.
+#[derive(Clone)]
+pub struct AesOcb3<Aes, NonceSize = U12, TagSize = U16>
+where
+    NonceSize: self::NonceSize,
+    TagSize: self::TagSize,
+{
+    cipher: Aes,
```

Comment:
This should probably just be `Ocb3` declared generic around a cipher `C`. I can fix that up after a merge.

Reply:
Also note `AesOcb3` can be achieved as a type alias so this all works out-of-the-box still.